File: src/main/java/at/yawk/fiction/Pageable.java
package at.yawk.fiction;

import java.util.List;
import lombok.Data;

/**
 * @author yawkat
 */
public interface Pageable<T> {
    /**
     * Get the page of the given index. Indexes start at 0.
     */
    Page<T> getPage(int i) throws Exception;

    @Data
    final class Page<T> {
        List<T> entries;
        /**
         * The expected total page count or <code>-1</code> if unknown.
         */
        int pageCount = -1;
        /**
         * <code>true</code> if this is the last page, <code>false</code> if there are more following pages or
         * we don't know.
         */
        boolean last = false;
    }
}

File: src/main/kotlin/at/yawk/fiction/Pageable.kt
package at.yawk.fiction

/**
 * @author yawkat
 */
interface Pageable<T> {
    /**
     * Get the page of the given index. Indexes start at 0.
     */
    fun getPage(page: Int): Page<T>

    data class Page<T>(
            val entries: List<T>,
            /**
             * The expected total page count or `null` if unknown.
             */
            val pageCount: Int?,
            /**
             * `true` if this is the last page, `false` if there are more following pages or
             * we don't know.
             */
            val last: Boolean
    )
}
